### PR TITLE
Go extractor - Add Support for packing nested go modules

### PR DESCRIPTION
**Pull request by BarakHc**
- [ ] All [tests](https://ci.appveyor.com/project/jfrog-ecosystem/build-info) passed. If this feature is not already covered by the tests, I added new tests.
-----

### CODE REVIEW

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+     * @return The normalized path
+     */
+    public static CharSequence normalizeSlashes(CharSequence path) {
+        if (path == null) {
+            return null;
+        }
+        return PATTERN_SLASHES.matcher(path).replaceAll("/");
+    }
+
+
```
Comment (eyalbe4) [inline build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java:53]: Redundant new-line here.

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check that the given CharSequence is neither <code>null</code> nor of length 0. Note: Will return
+     * <code>true</code> for a CharSequence that purely consists of whitespace.
+     */
+    public static boolean hasLength(String str) {
```
Comment (eyalbe4) [inline build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java:75]: Why do we need this? ```StringUtils.isNotEmpty``` does this.

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        return PATTERN_SLASHES.matcher(path).replaceAll("/");
+    }
+
+
+    /**
+     * Check whether the given CharSequence has actual text. More specifically, returns <code>true</code> if the string
+     * not <code>null</code>, its length is greater than 0, and it contains at least one non-whitespace character.
+     */
+    public static boolean hasText(String str) {
```
Comment (eyalbe4) [inline build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java:58]: I think this implementation can be simplified - 
Let's just replace  all " " with "" and return if length > 0.

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * Replaces all backslashes in the path to slashes.
+     *
+     * @param path A path to format
+     * @return The input path with all backslashes replaced with slashes. Return empty string if the input path is
+     * null.
+     */
+    public static String formatPath(String path) {
+        if (hasText(path)) {
```
Comment (eyalbe4) [inline build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java:33]: This if seems redundant. What would be the problem if path does not have text?

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoZipBallStreamer.java [hunk 5]
```diff
@@ -102,6 +103,7 @@ private void initiateProjectType() {
             if (shouldPackSubModule()) {
                 log.debug(projectName + "@" + version + " is a sub module - the sub module name is " + subModuleName);
             } else {
+                subModuleName = StringUtils.EMPTY;
                 log.debug(projectName + "@" + version + " is a regular module");
             }
         }
```
[the remark this answers is not in the document]
Comment (BarakHc) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoZipBallStreamer.java]: The                 
_subModuleName = GoVersionUtils.getSubModule(projectName);_
Is actually necessary to determine  the result of _shouldPackSubModule()_.

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java (hunk 2 above)
Comment (BarakHc) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java]: removed

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java (hunk 3 above)
Comment (BarakHc) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java]: removed

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java (hunk 4 above)
Comment (BarakHc) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/PathUtils.java]: removed

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoVersionUtils.java [hunk 6]
```diff
@@ -97,12 +98,27 @@ public static boolean isCompatibleGoModuleNaming(String projectName, String vers
...
+     */
+    public static String getParent(String path) {
+        if (StringUtils.isEmpty(path)) {
+            return StringUtils.EMPTY;
+        }
+        String parentPath = new File(path).getParent();
+        if (StringUtils.isNotEmpty(parentPath)) {
+            return parentPath.replace('\\', '/');
+        }
+        return StringUtils.EMPTY;
```
Comment (eyalbe4) [inline build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoVersionUtils.java:122]: The above if statement seems to be redundant. The following code:
```
        if (StringUtils.isNotEmpty(parentPath)) {
            return parentPath.replace('\\', '/');
        }
        return StringUtils.EMPTY;
```
can be reduced to:
```
return parentPath.replace('\\', '/');
```

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoZipBallStreamer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (eyalbe4) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoZipBallStreamer.java]: I see. I didn't notice it is a class member...

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
See my above inline comment.

File: build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoVersionUtils.java (hunk 6 above)
Comment (BarakHc) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoVersionUtils.java:122]: But parentPath can be null

Comment (eyalbe4) [reply build-info-extractor-go/src/main/java/org/jfrog/build/extractor/go/extractor/GoVersionUtils.java:122]: You're correct!

**Pull request merged**